src/sender_queue/mod.rs

```diff
@@ -0,0 +1,366 @@
+//! # Sender queue
+//!
+//! A sender queue allows a `DistAlgorithm` that outputs `Epoched` messages to buffer those outgoing
+//! messages based on their epochs. A message is sent to its recipient only when the recipient's
+//! epoch matches the epoch of the message. Thus no queueing is required for incoming messages since
+//! any incoming messages with non-matching epochs can be safely discarded.
+
+mod message;
+
+use std::cmp::Ordering;
+use std::collections::{btree_map::Entry, BTreeMap, BTreeSet};
+use std::fmt::Debug;
+
+use rand::Rand;
+use serde::{de::DeserializeOwned, Serialize};
+
+use {DistAlgorithm, Epoched, NodeIdT, Target, TargetedMessage};
+
+pub use self::message::{Message, MessageContent};
+
+/// Generic sender queue functionality.
+pub trait SenderQueueFunc<D>
+where
+    D: DistAlgorithm,
+    <D as DistAlgorithm>::Message: Clone + Epoched + Serialize + DeserializeOwned,
+    D::NodeId: NodeIdT + Rand,
+{
+    type Step;
+
+    /// Returns a pair containing
+    ///
+    /// - the maximum epoch out of a given epoch and the epoch of a given batch,
+    ///
+    /// - an optional new peer to be added to the set of peers.
+    fn max_epoch_with_batch(
+        &self,
+        epoch: <D::Message as Epoched>::Epoch,
+        batch: &D::Output,
+    ) -> (<D::Message as Epoched>::Epoch, Option<D::NodeId>);
+
+    /// Whether the epoch `them` accepts the message `us`.
+    fn is_accepting_epoch(&self, us: &D::Message, them: <D::Message as Epoched>::Epoch) -> bool;
+
+    /// Whether the epoch `them` is ahead of the epoch of the message `us`.
+    fn is_later_epoch(&self, us: &D::Message, them: <D::Message as Epoched>::Epoch) -> bool;
```

Comment:
Maybe `is_message_obsolete`? Otherwise I'm never sure whether the message's or the peer's epoch is the "later" one.